go/vt/discovery/healthcheck.go

```diff
@@ -277,17 +277,37 @@ func (hc *HealthCheckImpl) servingConnStats() map[string]int64 {
 	return res
 }
 
+// finalizeConn closes the health checking connection and sends the final
+// notification about the tablet to downstream.
+func (hc *HealthCheckImpl) finalizeConn(hcc *healthCheckConn) {
+	hcc.mu.Lock()
+	if hcc.conn != nil {
+		hcc.conn.Close()
+		hcc.conn = nil
+	}
+	hcc.tabletStats.Up = false
+	hcc.tabletStats.Serving = false
+	hcc.tabletStats.LastError = hcc.ctx.Err()
```

Comment:
Note: `ctx.Err()` only returns a valid error if the channel returned by `ctx.Done()` is closed. See the documentation: https://golang.org/pkg/context/

We ran into problems before where `Err()` returned an undefined value (because the channel was not closed) and the value was not `nil`.

You can check the channel with a non-blocking read (https://gobyexample.com/non-blocking-channel-operations) and set the error only then.

I acknowledge that this a corner case which we can probably ignore in practice because the connection is closed anyway. But I would prefer if we use `ctx.Err()` correctly throughout the codebase :)